File: apps/android/app/src/test/java/com/litter/android/SessionsDerivationTests.kt
package com.litter.android

import com.litter.android.ui.sessions.SessionsDerivation
import com.litter.android.ui.sessions.WorkspaceSortMode
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import uniffi.codex_mobile_client.AppSessionSummary
import uniffi.codex_mobile_client.AppSubagentStatus
import uniffi.codex_mobile_client.ThreadKey

class SessionsDerivationTests {

    @Test
    fun `normalizedCwd trims trailing slash and lowercases`() {
        assertEquals("/home/user/projects", SessionsDerivation.normalizedCwd("/home/user/projects/"))
        assertEquals("/home/user/projects", SessionsDerivation.normalizedCwd("/home/user/Projects"))
    }

    @Test
    fun `normalizedCwd returns tilde for null or blank`() {
        assertEquals("~", SessionsDerivation.normalizedCwd(null))
        assertEquals("~", SessionsDerivation.normalizedCwd(""))
    }

    @Test
    fun `derive nests a local studio fork with its parent and preserves the workspace group`() {
        val parent = session(threadId = "desktop-thread", updatedAt = 100, cwd = "/Projects/Litter")
        val child = session(
            threadId = "phone-fork",
            updatedAt = 200,
            cwd = "/Projects/Litter",
            parentThreadId = parent.key.threadId,
            isFork = true,
        )

        val result = SessionsDerivation.derive(listOf(child, parent))

        assertEquals(2, result.totalCount)
        assertEquals(1, result.groups.size)
        assertEquals("studio|/projects/litter", result.workspaceGroupKeys.single())
        assertEquals("desktop-thread", result.groups.single().nodes.single().summary.key.threadId)
        assertEquals("phone-fork", result.groups.single().nodes.single().children.single().summary.key.threadId)
        assertEquals(parent.key, result.parentByKey.getValue(child.key).key)
        assertEquals("studio|/projects/litter", result.workspaceGroupKeyByThreadKey.getValue(child.key))
    }

    @Test
    fun `derive makes a filtered-out parent a root instead of hiding the child`() {
        val parent = session(threadId = "parent", updatedAt = 100, title = "Desktop work")
        val child = session(
            threadId = "child",
            updatedAt = 200,
            title = "Phone follow-up",
            parentThreadId = parent.key.threadId,
            isFork = true,
        )

        val result = SessionsDerivation.derive(
            summaries = listOf(parent, child),
            forkOnly = true,
        )

        assertEquals(listOf(child.key), result.filteredThreadKeys)
        assertEquals("child", result.groups.single().nodes.single().summary.key.threadId)
        assertTrue(result.groups.single().nodes.single().children.isEmpty())
        assertTrue(result.parentByKey.isEmpty())
    }

    @Test
    fun `derive search matches Local Studio metadata and sorts workspace names`() {
        val localStudio = session(
            threadId = "studio",
            updatedAt = 100,
            cwd = "/work/Zebra",
            agentDisplayLabel = "Local Studio",
        )
        val codex = session(
            serverId = "codex",
            threadId = "codex",
            updatedAt = 200,
            cwd = "/work/alpha",
            title = "Other work",
        )

        val searched = SessionsDerivation.derive(
            summaries = listOf(codex, localStudio),
            searchQuery = "studio",
        )
        assertEquals(listOf(localStudio.key), searched.filteredThreadKeys)

        val alphabetical = SessionsDerivation.derive(
            summaries = listOf(localStudio, codex),
            sortMode = WorkspaceSortMode.NAME,
        )
        assertEquals(listOf("alpha", "Zebra"), alphabetical.groups.map { it.workspaceLabel })
    }

    private fun session(
        serverId: String = "studio",
        threadId: String,
        updatedAt: Long,
        cwd: String = "/work/litter",
        title: String = threadId,
        parentThreadId: String? = null,
        isFork: Boolean = false,
        agentDisplayLabel: String? = null,
    ) = AppSessionSummary(
        key = ThreadKey(serverId = serverId, threadId = threadId),
        agentRuntimeKind = if (serverId == "studio") "local-studio" else "codex",
        serverDisplayName = if (serverId == "studio") "Local Studio" else "Codex",
        serverHost = "$serverId.local",
        title = title,
        preview = title,
        cwd = cwd,
        model = "",
        modelProvider = "",
        parentThreadId = parentThreadId,
        forkedFromId = parentThreadId,
        agentNickname = null,
        agentRole = null,
        agentDisplayLabel = agentDisplayLabel,
        agentStatus = AppSubagentStatus.UNKNOWN,
        updatedAt = updatedAt,
        hasActiveTurn = false,
        isResumed = false,
        isSubagent = false,
        isFork = isFork,
        lastResponsePreview = null,
        lastResponseTurnId = null,
        lastUserMessage = null,
        lastToolLabel = null,
        recentToolLog = emptyList(),
        lastTurnStartMs = null,
        lastTurnEndMs = null,
        stats = null,
        tokenUsage = null,
        goal = null,
    )
}

File: apps/ios/Tests/LitterTests/SessionsDerivationTests.swift
import XCTest
@testable import Litter

@MainActor
final class SessionsDerivationTests: XCTestCase {
    func testSiblingsKeepServerScopeAndMostRecentOrder() {
        let sessions = [
            summary("parent"),
            summary("selected", parent: " parent ", updatedAt: 20),
            summary("older", parent: "parent", updatedAt: 10),
            summary("newer", parent: "parent", updatedAt: 30),
            summary("parent", server: "other"),
            summary("foreign", server: "other", parent: "parent", updatedAt: 40)
        ]
        let derived = build(sessions)
        XCTAssertEqual(derived.siblings(for: sessions[1].key).map(\.threadId), ["newer", "older"])
        XCTAssertTrue(derived.siblings(for: sessions[0].key).isEmpty)
        XCTAssertTrue(derived.siblings(for: sessions[5].key).isEmpty)
    }

    func testMissingParentDoesNotInventSiblingNavigation() {
        let sessions = [summary("one", parent: "missing"), summary("two", parent: "missing")]
        let derived = build(sessions)
        XCTAssertTrue(derived.siblings(for: sessions[0].key).isEmpty)
        XCTAssertEqual(derived.workspaceSections.first?.groups.first?.treeRoots.count, 2)
    }

    func testFilteredListRetainsNavigationToHiddenSiblings() {
        let sessions = [summary("parent"), summary("selected", parent: "parent"), summary("hidden", parent: "parent")]
        let derived = build(sessions, searchQuery: "selected")
        XCTAssertEqual(derived.filteredThreadKeys, [sessions[1].key])
        XCTAssertEqual(derived.siblings(for: sessions[1].key).map(\.threadId), ["hidden"])
    }

    func testLargeSiblingFamiliesRetainOnlyOneChildList() {
        for count in [100, 1_000] {
            let sessions = [summary("parent")] + (0..<count).map {
                summary("child-\($0)", parent: "parent", updatedAt: Int64($0))
            }
            let start = ContinuousClock.now
            let derived = build(sessions)
            let elapsed = start.duration(to: .now)
            XCTAssertEqual(derived.childrenByKey.values.reduce(0) { $0 + $1.count }, count)
            XCTAssertEqual(derived.siblings(for: sessions[1].key).count, count - 1)
            XCTAssertEqual(derived.allThreadKeys.count, count + 1)
            print("SessionsDerivation siblings=\(count) elapsed=\(elapsed) retainedChildEntries=\(count)")
        }
    }

    private func build(_ sessions: [AppSessionSummary], searchQuery: String = "") -> SessionsDerivedData {
        SessionsDerivation.build(
            sessions: sessions,
            selectedServerFilterId: nil,
            showOnlyForks: false,
            selectedRuntimeKind: nil,
            workspaceSortMode: .mostRecent,
            searchQuery: searchQuery,
            frozenMostRecentOrder: nil
        )
    }

    private func summary(
        _ id: String,
        server: String = "server",
        parent: String? = nil,
        updatedAt: Int64 = 0
    ) -> AppSessionSummary {
        AppSessionSummary(
            key: ThreadKey(serverId: server, threadId: id),
            agentRuntimeKind: .codex,
            serverDisplayName: server,
            serverHost: "\(server).local",
            title: id,
            preview: "",
            cwd: "/workspace",
            model: "",
            modelProvider: "",
            parentThreadId: parent,
            forkedFromId: nil,
            agentNickname: nil,
            agentRole: nil,
            agentDisplayLabel: nil,
            agentStatus: .unknown,
            updatedAt: updatedAt,
            hasActiveTurn: false,
            isResumed: false,
            isSubagent: parent != nil,
            isFork: parent != nil,
            lastResponsePreview: nil,
            lastResponseTurnId: nil,
            lastUserMessage: nil,
            lastToolLabel: nil,
            recentToolLog: [],
            lastTurnStartMs: nil,
            lastTurnEndMs: nil,
            stats: nil,
            tokenUsage: nil,
            goal: nil
        )
    }
}
